File: java/src/test/java/io/github/ajoz/validation/SimpleTest.java
package io.github.ajoz.validation;

import org.junit.Assert;
import org.junit.Test;

import java.util.function.Function;

public class SimpleTest {
    private final static Function<Integer, Function<Integer, Integer>> plus =
            first -> (Function<Integer, Integer>) second -> first + second;

    @Test
    public void shouldSumTwoValues() {
        final Validation<ErrorMessage, Function<Integer, Function<Integer, Integer>>> first = Validation.success(plus);
        final Validation<ErrorMessage, Integer> second = Validation.success(1);
        final Validation<ErrorMessage, Integer> third = Validation.success(2);

        final Validation<ErrorMessage, Integer> combined = Validation.ap(Validation.ap(first, second), third);
        final int value = combined.getValue();
        Assert.assertEquals(3, value);
    }
}

File: kotlin/src/test/kotlin/io/github/ajoz/validation/SimpleTest.kt
package io.github.ajoz.validation

import io.github.ajoz.curry
import io.github.ajoz.validation.Validation.Success
import org.junit.Assert.assertEquals
import org.junit.Test

typealias IntOptSuccess = Success<ErrorMessage, (Int) -> (Int) -> Int>
typealias IntSuccess = Success<ErrorMessage, Int>

class SimpleTest {

    @Test
    fun shouldSumTwoValues() {
        val first = IntOptSuccess(curry(::plus))
        val second = IntSuccess(1)
        val third = IntSuccess(2)

        val combined = first ap second ap third
        val value = combined.value
        assertEquals(3, value.toLong())
    }
}

fun plus(a: Int, b: Int) = a + b